numpy/core/src/multiarray/iterators.c

```diff
@@ -640,34 +640,34 @@ iter_subscript(PyArrayIterObject *self, PyObject *ind)
         obj = ind;
     }
 
-    if (PyArray_Check(obj)) {
-        /* Check for Boolean object */
-        if (PyArray_TYPE((PyArrayObject *)obj) == NPY_BOOL) {
-            ret = iter_subscript_Bool(self, (PyArrayObject *)obj);
-            Py_DECREF(indtype);
-        }
-        /* Check for integer array */
-        else if (PyArray_ISINTEGER((PyArrayObject *)obj)) {
-            PyObject *new;
-            new = PyArray_FromAny(obj, indtype, 0, 0,
-                              NPY_ARRAY_FORCECAST | NPY_ARRAY_ALIGNED, NULL);
-            if (new == NULL) {
-                goto fail;
-            }
-            Py_DECREF(obj);
-            obj = new;
-            new = iter_subscript_int(self, (PyArrayObject *)obj);
-            Py_DECREF(obj);
-            return new;
-        }
-        else {
-            goto fail;
-        }
+    /* Any remaining valid input is an array or has been turned into one */
+    if (!PyArray_Check(obj)) {
+        goto fail;
+    }
+
+    /* Check for Boolean array */
+    if (PyArray_TYPE((PyArrayObject *)obj) == NPY_BOOL) {
+        ret = iter_subscript_Bool(self, (PyArrayObject *)obj);
+        Py_DECREF(indtype);
         Py_DECREF(obj);
         return (PyObject *)ret;
     }
-    else {
+
+    /* Check for integer array */
+    if (PyArray_ISINTEGER((PyArrayObject *)obj)) {
```

Comment:
Can we still invert this? I.e., 
```
if (!PyArray_ISINTEGER(...) {
    goto fail;
}
```
That way we don't use `fail` as a fall-through at all.

Reply:
I don't think that's equivalent. The `else` of this if is a no-op. The only remaining fall-through is on line 662 when the conversion to an array failed. Similar to the fall-through on line 634. Am I missing something?

Reply:
I mean, we can invert that `if` and move the contents of the original `if` block up one indent, but we'd still have that one `goto fail` left. I thought your remark was in order to remove the latter.

Reply:
No, that's not what I meant. Right now, you have *here*
```
if (PyArra...) {
  ...
  (goto fail or return something)
}

fail:
```

I'd like it to be
```
if (!PyArray_IS...) {
   goto fail;
}

...
(either goto fail or return something)

(No way to get here except via goto)
fail:
```

Just the same as for the original `if` statement!

Reply:
Though I guess the one issue is that you are creating `new` inside here, which is not possible if we unindent it.  So, perhaps just let it be...

Reply:
Ah, I see! I did understand what you were suggesting, but I thought that would be equivalent to what we have now. I somehow managed to forget about the (implicitly no-op) `else` branch of that `if` block... I'm sorry, you're right of course.

Is there any reason why we can't just move the declaration of `new` up next to that of `obj`, for instance? It took me a while to figure out why merely unindenting would not be possible (i.e. that we need to support ANSI C which restricts where declarations can be put) so I don't trust my judgment very much at this hour :)

Reply:
@adeak - indeed, one can just define `new` at the top - which some don't like to the point of adding clauses without even an `if` statement and others do not mind at all. Since you are doing the work, I think you should pick what you prefer! Though we should not let this distract from the fact that you solved the bug! So, if you'd rather just have us have a final look so that this can be merged, that's totally OK!

Reply:
Perfect, thanks for the explanation, @mhvk. I did realize the diabolical option to add a superfluous clause. Since the three options seem to be "early declaration", "superfluous clause" and "fallthrough into `fail`", I find the first to be the least offensive. And if by any chance it's decided that my choice is worse after all, whoever merges can just cherry-pick out that commit and keep the current state.